feat: Login, Sign-Up, Authentication

TheRealImaginary opened the pull request.

melzareix committed:
feat: Routing boilerplate

melzareix committed:
Merge branch 'master' of https://github.com/melzareix/voting-app into user-authentication

melzareix committed:
feat: Initial Signup route.

TheRealImaginary committed:
feat: Added Login, Signup Routes and Authentication, Error Handling Middleware

Added routes for login and signup. Moreover added Error Handling and Authentication Middlewares. Found and fixed an error/bug in the Models.

TheRealImaginary committed:
style: formatting, parenthesis;no code change

melzareix commented:
### Changes Required
- Include error handling middleware(s) for apiV1.
- Refactor '/secret' to use a general authentication middleware.
- Refactor User model to use email instead of name and add server side email format verification.
- Unify code by using ``` return ``` before ```res.json``` and ```next()```.
- Use ```const``` or ```let``` appropriately instead of ```var```.

TheRealImaginary committed:
refactor: Refactored code after it was reviewed

-Error handling middleware is now in apiV1.
-Authentication middleware for '/secret' refactored.
-Refactored User model to use email instead of name and added server side email verification.
-Unified code style and added return before sending a response/using next middleware.
-Using ES6 const and let instead of var.

TheRealImaginary committed:
refactor: refactored /secret authentication middleware

melzareix commented on app.js:
```diff
@@ -14,21 +14,6 @@ app.get('/', function (req, res) {
 	res.end('Hello World.');
 });
 
```
Add a generic middleware.

melzareix replied:
Add a generic middleware.

melzareix commented on routes/apiv1.js:
```diff
@@ -22,33 +22,33 @@ router.use(bodyParser.json());
 		return next();
 	User.findOne({
-		name: username
+		email: email
```
We are using ES6, so I think sending email only
```
return {
   email
};
```
is better.

melzareix commented on routes/apiv1.js:
```diff
@@ -76,13 +76,27 @@ router.post('/login', function (req, res, next) {
 	});
 });
 
+//TODO : Pass new Error to next()
```
Remove ```TODO``` when finished.

melzareix commented on models/User.js:
```diff
@@ -20,6 +21,10 @@ const userSchema = mongoose.Schema({
 userSchema.pre('save', function (done) {
     var user = this;
 
+    if (!validator.isEmail(user.email)) {
```
Use Mongoose Validation for a cleaner code
[https://github.com/leepowellcouk/mongoose-validator](https://github.com/leepowellcouk/mongoose-validator)

melzareix commented on routes/apiv1.js:
```diff
@@ -22,33 +22,33 @@ router.use(bodyParser.json());
 	});
 	user.save(function (err) {
-		//Has Duplicate then Error
+		//Has Duplicate OR Invalid Email
```
Use comment as regular english
```
Has duplicate or invalid email.
```

TheRealImaginary committed:
refactor: Refactored code as per review

melzareix reviewed (changes requested):
Good Job!
Need only to return unauthorized page as JSON!

TheRealImaginary committed:
refactor: Refactored code and made custom JSON un-authorization

melzareix commented on routes/apiv1.js:
```diff
@@ -76,9 +76,19 @@ router.post('/login', function (req, res, next) {
+router.use(function (req, res, next) {
+	passport.authenticate('jwt', {
+		session: false
+	}, function (err, user, info) {
```
Passport is kinda stupid or we are but anyway,
1 - err is not null only in case of an internal error or something or with mongo for example, you define this in authenticator line 17.
2 - Errors from JWT passport are passed as error objects in `info`.
3 - if `JWT` token is invalid info will contain an error raised from the strategy, the error is not serializable thus you need to use toString on it.
4 - `Invalid Credentials` will occur only if the user is not found in the database, so you need to pass it as `new error()` to be like point 3.

TheRealImaginary committed:
refactor: Refactored code per review

TheRealImaginary committed:
style: Added New line

melzareix commented:
Resolves #12 .
